internal/explorer/server.go

```diff
@@ -28,56 +29,55 @@ import (
 // @Router /farms [get]
 func (a *App) listFarms(w http.ResponseWriter, r *http.Request) {
 	enableCors(&w)
-	r, err := a.handleRequestsQueryParams(r)
+	filter, limit, err := a.handleFarmRequestsQueryParams(r)
 	if err != nil {
 		errorReplyWithStatus(err, w, http.StatusBadRequest)
 		return
 	}
-	maxResult, pageOffset := getMaxResult(r.Context()), getOffset(r.Context())
-
-	queryString := fmt.Sprintf(`
-	{
-		farms (limit:%d,offset:%d) {
-			name
-			farmId
-			twinId
-			version
-			farmId
-			pricingPolicyId
-			stellarAddress
-			publicIPs{
-				id
-				ip
-				contractId
-				gateway
-			}
-		}
-	}
-	`, maxResult, pageOffset)
-
-	farms := FarmResult{}
-	err = a.query(queryString, &farms)
-
+	farms, err := a.db.GetFarms(filter, limit)
 	if err != nil {
 		log.Error().Err(err).Msg("failed to query farm")
 		errorReplyWithStatus(err, w, http.StatusInternalServerError)
 		return
 	}
-
-	result, err := json.Marshal(farms.Data.Farms)
+	serialzied, err := json.Marshal(farms)
 	if err != nil {
 		log.Error().Err(err).Msg("failed to marshal farm")
 		errorReplyWithStatus(err, w, http.StatusInternalServerError)
 		return
 	}
 	w.Header().Add("Content-Type", "application/json")
 	w.WriteHeader(http.StatusOK)
-	w.Write([]byte(result))
+	w.Write(serialzied)
```

Comment:
the http handlers were built to be generic this is why u can write any header, or data to the output stream. But apps can still build some middleware layers to suit their functionality. Hence I once created a util that make this easier for our use case (i probably need to create library out of it) but for now u can just copy it.

I am not saying it's perfect but it will make your life in http handler much much easier. For now just copy it from here `https://github.com/threefoldtech/zos/blob/main/pkg/provision/mw/action.go`

then you can do 

```go
router.HandleFunc("/farms", AsActionHandler(a.listFarms))
```
then list farm signature can change according. inside the code you can  `return result, nil`, or one of the other responses like `return nil, NotFound(err)` or build your own response. But then u don't care about serialization, nor writing headers, etc. ...